src/chabochethermal.jl

```diff
@@ -0,0 +1,1050 @@
+# This file is a part of JuliaFEM.
+# License is MIT: see https://github.com/JuliaFEM/Materials.jl/blob/master/LICENSE
+
+module ChabocheThermalModule
+
+using LinearAlgebra, ForwardDiff, Tensors, NLsolve, Parameters
+
+import ..AbstractMaterial, ..AbstractMaterialState
+import ..Utilities: Symm2, Symm4, isotropic_elasticity_tensor, isotropic_compliance_tensor, lame, debang
+import ..integrate_material!  # for method extension
+
+# parametrically polymorphic for any type representing ℝ
+export GenericChabocheThermal, GenericChabocheThermalDriverState, GenericChabocheThermalParameterState, GenericChabocheThermalVariableState
+
+# specialization for Float64
+export ChabocheThermal, ChabocheThermalDriverState, ChabocheThermalParameterState, ChabocheThermalVariableState
+
+"""Rank-2 identity tensor in three spatial dimensions."""
+I2 = Symm2(I(3))
+
+@with_kw mutable struct GenericChabocheThermalDriverState{T <: Real} <: AbstractMaterialState
+    time::T = zero(T)
+    strain::Symm2{T} = zero(Symm2{T})
+    temperature::T = zero(T)
+end
+
+# TODO: hierarchize parameters: elasticity, kinematic hardening, isotropic hardening, ...
+# plasticity: yield criterion, flow rule, hardening
+"""Parameter state for ChabocheThermal material.
+
+The classical viscoplastic material is a special case of this model with `C1 = C2 = C3 = 0`.
+
+Maximum hardening for each backstress is `Cj / Dj`.
+
+Any parameter that is a `Function` should takes a single argument, the absolute
```

Comment:
I had a similar design in my other project. However, should the `Interpolations`-package interpolation objects be considered (in the future, more modularized structure)? This way we could have better control for the type of interpolation (1D, 2D, etc) from the types of interpolation objects. E.g. the Elastic modulus of wood is a function of temperature and humidity.

Reply:
Perhaps. Would you happen to know if `Interpolations` is autodiffable by `ForwardDiff`? By default I would expect so, but quickly skimming the documentation, I didn't find a mention of this. The docs did mention that some interpolation algorithms can provide jacobians and hessians.

It's important because we obtain quantities such as *n = ∂f/∂σ* and the algorithmic jacobian *∂σ/∂ε* by autodiffing - so the whole code path must be differentiable. (Twice, if *n* appears in the algorithmic jacobian.)

Reply:
I don't know, but I agree it is vital.

Reply:
Ok, I suppose we'll just test it at some point. :)